# 프로젝트 설명(교내 학술제)

아래 코드는 프로젝트 진행 과정 및 AI/머신러닝 기반 데이터 분석을 수행하는 과정으로 구성되어 있습니다.


주제 : AI데이터분석 분야: 미래를 예측하는 데이터 챌린지 – 우리 동네 가맹점, 위기 신호를 미리 잡아라! (LLM 활용 영세/중소 요식 가맹점 경영 위기 조기 경보 시스템 제안)

범위: 서울 성동구 (4,185건의 데이터 : 2023-2024 2개년 데이터)

[빅데이터 콘테스트](https://www.bigcontest.or.kr/)
[프로젝트 주요 정리 노션](https://www.notion.so/27f796ab4c2d808fb376e73b2192aec3?source=copy_link)

📌 주요 내용  
- 데이터 전처리 및 피처 정제  
- 각 데이터 개월수 별 모델 평가(2,3,4,5개월)
- 모델 학습  
- 모델 평가  
- LLM 활용 챗봇

작성자: 김한세  
작성일: 2025-11-23


In [ ]:
# 데이터 불러오기
import pandas as pd

df2 = pd.read_csv("big_data_set2_f.csv", encoding="cp949")

df2

ENCODED_MCT   TA_YM       MCT_OPE_MS_CN           RC_M1_SAA  \
0      000F03E44A  202404            4_50-75%            5_75-90%   
1      000F03E44A  202312            4_50-75%  6_90%초과(하위 10% 이하)   
2      002816BA73  202404            2_10-25%            3_25-50%   
3      002816BA73  202411            2_10-25%            3_25-50%   
4      002816BA73  202406            2_10-25%            4_50-75%   
...           ...     ...                 ...                 ...   
86585  A2D9C941A9  202304            5_75-90%            2_10-25%   
86586  A7C2307794  202303            2_10-25%            2_10-25%   
86587  BF6A6456AB  202407  6_90%초과(하위 10% 이하)            2_10-25%   
86588  C59F032CAC  202406             1_10%이하            2_10-25%   
86589  F46D45E328  202302            2_10-25%            2_10-25%   

           RC_M1_TO_UE_CT     RC_M1_UE_CUS_CN      RC_M1_AV_NP_AT  \
0                5_75-90%            5_75-90%            5_75-90%   
1      6_90%초과(하위 10% 이하)  6_90%초과(하위 10% 이하)  6_90%초과(하위 10% 이하)   
2                4_50-75%            4_50-75%            2_10-25%   
3                4_50-75%            4_50-75%            2_10-25%   
4                4_50-75%            4_50-75%            2_10-25%   
...                   ...                 ...                 ...   
86585             1_10%이하             1_10%이하  6_90%초과(하위 10% 이하)   
86586            2_10-25%            2_10-25%            3_25-50%   
86587            4_50-75%            4_50-75%             1_10%이하   
86588            2_10-25%            2_10-25%            4_50-75%   
86589            3_25-50%            3_25-50%            3_25-50%   

           APV_CE_RAT  DLV_SAA_RAT  M1_SME_RY_SAA_RAT  M1_SME_RY_CNT_RAT  \
0             1_상위1구간    -999999.9                2.6               10.6   
1                 NaN    -999999.9                0.0                0.0   
2      6_상위6구간(하위1구간)    -999999.9               96.6               40.8   
3      6_상위6구간(하위1구간)    -999999.9              108.0               46.4   
4             1_상위1구간    -999999.9               74.2               38.1   
...               ...          ...                ...                ...   
86585         2_상위2구간          2.6              249.7              465.0   
86586         1_상위1구간         19.2              196.7              198.8   
86587  6_상위6구간(하위1구간)          1.6              176.3               48.7   
86588         2_상위2구간         24.0              116.5               74.5   
86589         1_상위1구간         12.8              215.6              139.7   

       M12_SME_RY_SAA_PCE_RT  M12_SME_BZN_SAA_PCE_RT  M12_SME_RY_ME_MCT_RAT  \
0                       93.8                    71.5                   16.7   
1                       94.8                    73.4                   16.6   
2                       15.6                    20.0                   17.5   
3                       16.8                    19.8                   16.9   
4                       16.0                    19.9                   17.3   
...                      ...                     ...                    ...   
86585                   10.2                     0.0                   11.4   
86586                   10.1                    14.1                   15.9   
86587                   11.5                    16.5                   19.1   
86588                   26.6                    30.2                   12.3   
86589                    6.2                     6.6                   15.5   

       M12_SME_BZN_ME_MCT_RAT  
0                         7.8  
1                         7.2  
2                         5.2  
3                         6.9  
4                         5.6  
...                       ...  
86585               -999999.9  
86586                     8.8  
86587                     8.7  
86588                    10.9  
86589                     9.2  

[86590 rows x 15 columns]

In [2]:
df1= pd.read_csv("big_data_set1_f.csv", encoding="cp949")

df1

ENCODED_MCT           MCT_BSE_AR        MCT_NM MCT_BRD_NUM  \
0     16184E93D9           서울 성동구 마장동          성우**         NaN   
1     4D039EA8B7           서울 성동구 마장동          대보**         NaN   
2     0074C4990A           서울 성동구 마장동          대용**         NaN   
3     68308F2746           서울 성동구 마장동          통일**         NaN   
4     4117EDDE9C           서울 성동구 마장동          한울**         NaN   
...          ...                  ...           ...         ...   
4180  7FCF23E6F3    서울특별시 성동구 마장로 137  아이**********         NaN   
4181  9743B78531      서울 성동구 독서당로 223          어반**         NaN   
4182  C2261977A8  서울특별시 성동구 마장로35길 76          한성**         NaN   
4183  F1C69918DF      서울 성동구 왕십리로 410          국수**   한식-국수/만두3   
4184  58B8C943BF     서울특별시 성동구 매봉길 50      아이******         NaN   

     MCT_SIGUNGU_NM HPSN_MCT_ZCD_NM HPSN_MCT_BZN_CD_NM     ARE_D  MCT_ME_D  
0            서울 성동구             축산물                마장동  20130320       NaN  
1            서울 성동구             축산물                마장동  20131122       NaN  
2            서울 성동구             축산물                마장동  20140512       NaN  
3            서울 성동구             축산물                마장동  20151124       NaN  
4            서울 성동구             축산물                마장동  20151211       NaN  
...             ...             ...                ...       ...       ...  
4180         서울 성동구        아이스크림/빙수                NaN  20230821       NaN  
4181         서울 성동구              양식                NaN  20130305       NaN  
4182         서울 성동구             축산물                NaN  20230517       NaN  
4183         서울 성동구        한식-국수/만두                NaN  20211007       NaN  
4184         서울 성동구        아이스크림/빙수                NaN  20230303       NaN  

[4185 rows x 9 columns]

In [ ]:
dong = pd.read_csv("big_data_with_hdong_final.csv")
data_dong = dong[['ENCODED_MCT','행정동']]

In [ ]:
# -----------------------------
# 1️⃣ df에서 필요한 컬럼만 추출
# -----------------------------
df_sel = df2[[
    'ENCODED_MCT',
    'TA_YM',
    'MCT_OPE_MS_CN',
    'RC_M1_SAA',
    'RC_M1_TO_UE_CT',
    'M1_SME_RY_SAA_RAT',
    'M1_SME_RY_CNT_RAT',
    'M12_SME_RY_SAA_PCE_RT',
    'M12_SME_BZN_SAA_PCE_RT',
    'M12_SME_RY_ME_MCT_RAT',
    'M12_SME_BZN_ME_MCT_RAT'
]].copy()

# -----------------------------
# 2️⃣ data에서 필요한 컬럼만 추출
# -----------------------------
data_sel = df1[[
    'ENCODED_MCT',
    'MCT_BSE_AR',    # 주소
    'HPSN_MCT_ZCD_NM',  # 업종
    'ARE_D',         # 개설일
    'MCT_ME_D'       # 폐업일
]].copy()

# -----------------------------
# 4️⃣ 두 데이터 병합 (가맹점 기준)
# -----------------------------
final_df = pd.merge(df_sel, data_sel, on='ENCODED_MCT', how='left')
final_df = pd.merge(final_df, data_dong, on='ENCODED_MCT', how='left')

# -----------------------------
# 5️⃣ 결과 확인
# -----------------------------
final_df

ENCODED_MCT   TA_YM       MCT_OPE_MS_CN           RC_M1_SAA  \
0      000F03E44A  202404            4_50-75%            5_75-90%   
1      000F03E44A  202312            4_50-75%  6_90%초과(하위 10% 이하)   
2      002816BA73  202404            2_10-25%            3_25-50%   
3      002816BA73  202411            2_10-25%            3_25-50%   
4      002816BA73  202406            2_10-25%            4_50-75%   
...           ...     ...                 ...                 ...   
86585  A2D9C941A9  202304            5_75-90%            2_10-25%   
86586  A7C2307794  202303            2_10-25%            2_10-25%   
86587  BF6A6456AB  202407  6_90%초과(하위 10% 이하)            2_10-25%   
86588  C59F032CAC  202406             1_10%이하            2_10-25%   
86589  F46D45E328  202302            2_10-25%            2_10-25%   

           RC_M1_TO_UE_CT  M1_SME_RY_SAA_RAT  M1_SME_RY_CNT_RAT  \
0                5_75-90%                2.6               10.6   
1      6_90%초과(하위 10% 이하)                0.0                0.0   
2                4_50-75%               96.6               40.8   
3                4_50-75%              108.0               46.4   
4                4_50-75%               74.2               38.1   
...                   ...                ...                ...   
86585             1_10%이하              249.7              465.0   
86586            2_10-25%              196.7              198.8   
86587            4_50-75%              176.3               48.7   
86588            2_10-25%              116.5               74.5   
86589            3_25-50%              215.6              139.7   

       M12_SME_RY_SAA_PCE_RT  M12_SME_BZN_SAA_PCE_RT  M12_SME_RY_ME_MCT_RAT  \
0                       93.8                    71.5                   16.7   
1                       94.8                    73.4                   16.6   
2                       15.6                    20.0                   17.5   
3                       16.8                    19.8                   16.9   
4                       16.0                    19.9                   17.3   
...                      ...                     ...                    ...   
86585                   10.2                     0.0                   11.4   
86586                   10.1                    14.1                   15.9   
86587                   11.5                    16.5                   19.1   
86588                   26.6                    30.2                   12.3   
86589                    6.2                     6.6                   15.5   

       M12_SME_BZN_ME_MCT_RAT           MCT_BSE_AR HPSN_MCT_ZCD_NM     ARE_D  \
0                         7.8  서울특별시 성동구 왕십리로4가길 9      중식-딤섬/중식만두  20220225   
1                         7.2  서울특별시 성동구 왕십리로4가길 9      중식-딤섬/중식만두  20220225   
2                         5.2   서울 성동구 청계천로10나길 78            요리주점  20140507   
3                         6.9   서울 성동구 청계천로10나길 78            요리주점  20140507   
4                         5.6   서울 성동구 청계천로10나길 78            요리주점  20140507   
...                       ...                  ...             ...       ...   
86585               -999999.9    서울특별시 성동구 행당로5길 2           커피전문점  20220823   
86586                     8.8      서울 성동구 한림말1길 8.         한식-두부요리  20140508   
86587                     8.7    서울특별시 성동구 상원4길 11        한식-육류/고기  20240131   
86588                    10.9    서울 성동구 왕십리로 223-1        샌드위치/토스트  20090421   
86589                     9.2   서울 성동구 용답중앙15길 13.             중식당  20131128   

         MCT_ME_D     행정동  
0             NaN  성수1가1동  
1             NaN  성수1가1동  
2             NaN     마장동  
3             NaN     마장동  
4             NaN     마장동  
...           ...     ...  
86585         NaN   금호1가동  
86586         NaN     옥수동  
86587  20250724.0  성수1가2동  
86588         NaN     사근동  
86589         NaN     용답동  

[86590 rows x 16 columns]

In [ ]:
# 업종 통합

import pandas as pd
import numpy as np

def get_large_category(sector):
    """
    세부 업종(HPSN_MCT_ZCD_NM) 문자열을 입력받아
    대분류 문자열을 반환합니다.
    """
    if pd.isna(sector):
        return np.nan 
    
    # 1. 한식
    if '한식-' in sector or sector in ['백반/가정식', '한정식', '기사식당', '한식뷔페', '구내식당/푸드코트']:
        return '한식'
    
    # 2. 카페/디저트
    if sector in ['카페', '커피전문점', '테이크아웃커피', '테마카페', '차', '주스',
                  '베이커리', '아이스크림/빙수', '도너츠', '떡/한과', '마카롱', '와플/크로플', '탕후루']:
        return '카페/디저트'
    
    # 3. 외식 (기타)
    if '일식-' in sector or sector in ['일식당', '양식', '중식당', '동남아/인도음식', '기타세계요리', '스테이크', 
                                   '중식-딤섬/중식만두', '중식-훠궈/마라탕']:
        return '외식(기타)'
    
    # 4. 간편식/분식
    if sector in ['분식', '치킨', '피자', '햄버거', '샌드위치/토스트', '도시락', '꼬치구이']:
        return '간편식/분식'
        
    # 5. 주점/유흥
    if sector in ['호프/맥주', '요리주점', '이자카야', '민속주점', '포장마차', '와인바', 
                  '일반 유흥주점', '룸살롱/단란주점']:
        return '주점/유흥'

    # 6. 소매/식재료
    if sector in ['축산물', '식료품', '농산물', '수산물', '청과물', '미곡상', '건어물', '유제품', '반찬']:
        return '소매/식재료'

    # 7. 기타/제조
    if sector in ['건강식품', '건강원', '인삼제품', '식품 제조', '떡/한과 제조', '담배', '와인샵', '주류']:
        return '기타/제조'

    return '기타(미분류)'

# --- 메인 코드 실행 ---

input_file = 'big_data_with_hdong_final.csv'
output_file = 'big_data_with_large_categories_FINAL.csv'

print("--- 스크립트 실행 시작 ---")

try:
    print(f"'{input_file}' 파일을 읽고 있습니다...")
    df = pd.read_csv(input_file)
    print("파일 읽기 완료.")

    if 'HPSN_MCT_ZCD_NM' not in df.columns:
        # 'HPSN_MCT_ZCD_NM' 컬럼이 없으면 여기서 KeyError를 발생시킴
        raise KeyError
    
    print(f"'HPSN_MCT_ZCD_NM' 컬럼을 대분류로 변환합니다...")
    df['HPSN_MCT_ZCD_NM'] = df['HPSN_MCT_ZCD_NM'].map(get_large_category)
    print("변환 완료.")

    df.to_csv(output_file, index=False, encoding='utf-8-sig')
    
    print(f"\n--- 작업 성공 ---")
    print(f"'{output_file}' 이름으로 파일 저장을 완료했습니다.")
    print("파일 내용을 확인해보세요.")

except FileNotFoundError:
    print(f"\n[오류] '{input_file}' 파일을 찾을 수 없습니다.")
    print("이 스크립트 파일과 원본 파일이 같은 폴더에 있는지 확인해주세요.")
    
except KeyError:
    # [오류 수정] as e: 부분을 제거하고 가장 단순한 구문으로 변경했습니다.
    print(f"\n[오류] 'HPSN_MCT_ZCD_NM' 컬럼을 찾지 못했습니다.")
    print("원본 파일에 해당 컬럼명이 정확히 존재하는지 확인해주세요.")
    
except Exception as e:
    print(f"\n[알 수 없는 오류] 작업 중 오류가 발생했습니다: {e}")

print("--- 스크립트 실행 종료 ---")

--- 스크립트 실행 시작 ---
'big_data_with_hdong_final.csv' 파일을 읽고 있습니다...
파일 읽기 완료.
'HPSN_MCT_ZCD_NM' 컬럼을 대분류로 변환합니다...
변환 완료.

--- 작업 성공 ---
'big_data_with_large_categories_FINAL.csv' 이름으로 파일 저장을 완료했습니다.
파일 내용을 확인해보세요.
--- 스크립트 실행 종료 ---


In [6]:
data_33 = pd.read_csv("big_data_with_large_categories_FINAL.csv")
data_3 = data_33[['ENCODED_MCT','HPSN_MCT_ZCD_NM']]
final_df = final_df.drop(columns=['HPSN_MCT_ZCD_NM'])
final_df = pd.merge(final_df, data_3 , on='ENCODED_MCT', how='left')
final_df

ENCODED_MCT   TA_YM       MCT_OPE_MS_CN           RC_M1_SAA  \
0      000F03E44A  202404            4_50-75%            5_75-90%   
1      000F03E44A  202312            4_50-75%  6_90%초과(하위 10% 이하)   
2      002816BA73  202404            2_10-25%            3_25-50%   
3      002816BA73  202411            2_10-25%            3_25-50%   
4      002816BA73  202406            2_10-25%            4_50-75%   
...           ...     ...                 ...                 ...   
86585  A2D9C941A9  202304            5_75-90%            2_10-25%   
86586  A7C2307794  202303            2_10-25%            2_10-25%   
86587  BF6A6456AB  202407  6_90%초과(하위 10% 이하)            2_10-25%   
86588  C59F032CAC  202406             1_10%이하            2_10-25%   
86589  F46D45E328  202302            2_10-25%            2_10-25%   

           RC_M1_TO_UE_CT  M1_SME_RY_SAA_RAT  M1_SME_RY_CNT_RAT  \
0                5_75-90%                2.6               10.6   
1      6_90%초과(하위 10% 이하)                0.0                0.0   
2                4_50-75%               96.6               40.8   
3                4_50-75%              108.0               46.4   
4                4_50-75%               74.2               38.1   
...                   ...                ...                ...   
86585             1_10%이하              249.7              465.0   
86586            2_10-25%              196.7              198.8   
86587            4_50-75%              176.3               48.7   
86588            2_10-25%              116.5               74.5   
86589            3_25-50%              215.6              139.7   

       M12_SME_RY_SAA_PCE_RT  M12_SME_BZN_SAA_PCE_RT  M12_SME_RY_ME_MCT_RAT  \
0                       93.8                    71.5                   16.7   
1                       94.8                    73.4                   16.6   
2                       15.6                    20.0                   17.5   
3                       16.8                    19.8                   16.9   
4                       16.0                    19.9                   17.3   
...                      ...                     ...                    ...   
86585                   10.2                     0.0                   11.4   
86586                   10.1                    14.1                   15.9   
86587                   11.5                    16.5                   19.1   
86588                   26.6                    30.2                   12.3   
86589                    6.2                     6.6                   15.5   

       M12_SME_BZN_ME_MCT_RAT           MCT_BSE_AR     ARE_D    MCT_ME_D  \
0                         7.8  서울특별시 성동구 왕십리로4가길 9  20220225         NaN   
1                         7.2  서울특별시 성동구 왕십리로4가길 9  20220225         NaN   
2                         5.2   서울 성동구 청계천로10나길 78  20140507         NaN   
3                         6.9   서울 성동구 청계천로10나길 78  20140507         NaN   
4                         5.6   서울 성동구 청계천로10나길 78  20140507         NaN   
...                       ...                  ...       ...         ...   
86585               -999999.9    서울특별시 성동구 행당로5길 2  20220823         NaN   
86586                     8.8      서울 성동구 한림말1길 8.  20140508         NaN   
86587                     8.7    서울특별시 성동구 상원4길 11  20240131  20250724.0   
86588                    10.9    서울 성동구 왕십리로 223-1  20090421         NaN   
86589                     9.2   서울 성동구 용답중앙15길 13.  20131128         NaN   

          행정동 HPSN_MCT_ZCD_NM  
0      성수1가1동          외식(기타)  
1      성수1가1동          외식(기타)  
2         마장동           주점/유흥  
3         마장동           주점/유흥  
4         마장동           주점/유흥  
...       ...             ...  
86585   금호1가동          카페/디저트  
86586     옥수동              한식  
86587  성수1가2동              한식  
86588     사근동          간편식/분식  
86589     용답동          외식(기타)  

[86590 rows x 16 columns]

In [ ]:
# 폐업 라벨 생성
# 폐업일이 존재하면 폐업한 것으로 판단
final_df['폐업여부'] = final_df["MCT_ME_D"].notna().astype(int)


n_months_df = (
    final_df.groupby('ENCODED_MCT')['TA_YM']
    .nunique()                               # 가맹점별 고유 기준년월 개수
    .reset_index()
    .rename(columns={'TA_YM': 'n_months'})   # 컬럼 이름 변경
)

# 원본 데이터프레임에 병합
final_df = pd.merge(final_df, n_months_df, on='ENCODED_MCT', how='left')


final_df

ENCODED_MCT   TA_YM       MCT_OPE_MS_CN           RC_M1_SAA  \
0      000F03E44A  202404            4_50-75%            5_75-90%   
1      000F03E44A  202312            4_50-75%  6_90%초과(하위 10% 이하)   
2      002816BA73  202404            2_10-25%            3_25-50%   
3      002816BA73  202411            2_10-25%            3_25-50%   
4      002816BA73  202406            2_10-25%            4_50-75%   
...           ...     ...                 ...                 ...   
86585  A2D9C941A9  202304            5_75-90%            2_10-25%   
86586  A7C2307794  202303            2_10-25%            2_10-25%   
86587  BF6A6456AB  202407  6_90%초과(하위 10% 이하)            2_10-25%   
86588  C59F032CAC  202406             1_10%이하            2_10-25%   
86589  F46D45E328  202302            2_10-25%            2_10-25%   

           RC_M1_TO_UE_CT  M1_SME_RY_SAA_RAT  M1_SME_RY_CNT_RAT  \
0                5_75-90%                2.6               10.6   
1      6_90%초과(하위 10% 이하)                0.0                0.0   
2                4_50-75%               96.6               40.8   
3                4_50-75%              108.0               46.4   
4                4_50-75%               74.2               38.1   
...                   ...                ...                ...   
86585             1_10%이하              249.7              465.0   
86586            2_10-25%              196.7              198.8   
86587            4_50-75%              176.3               48.7   
86588            2_10-25%              116.5               74.5   
86589            3_25-50%              215.6              139.7   

       M12_SME_RY_SAA_PCE_RT  M12_SME_BZN_SAA_PCE_RT  M12_SME_RY_ME_MCT_RAT  \
0                       93.8                    71.5                   16.7   
1                       94.8                    73.4                   16.6   
2                       15.6                    20.0                   17.5   
3                       16.8                    19.8                   16.9   
4                       16.0                    19.9                   17.3   
...                      ...                     ...                    ...   
86585                   10.2                     0.0                   11.4   
86586                   10.1                    14.1                   15.9   
86587                   11.5                    16.5                   19.1   
86588                   26.6                    30.2                   12.3   
86589                    6.2                     6.6                   15.5   

       M12_SME_BZN_ME_MCT_RAT           MCT_BSE_AR     ARE_D    MCT_ME_D  \
0                         7.8  서울특별시 성동구 왕십리로4가길 9  20220225         NaN   
1                         7.2  서울특별시 성동구 왕십리로4가길 9  20220225         NaN   
2                         5.2   서울 성동구 청계천로10나길 78  20140507         NaN   
3                         6.9   서울 성동구 청계천로10나길 78  20140507         NaN   
4                         5.6   서울 성동구 청계천로10나길 78  20140507         NaN   
...                       ...                  ...       ...         ...   
86585               -999999.9    서울특별시 성동구 행당로5길 2  20220823         NaN   
86586                     8.8      서울 성동구 한림말1길 8.  20140508         NaN   
86587                     8.7    서울특별시 성동구 상원4길 11  20240131  20250724.0   
86588                    10.9    서울 성동구 왕십리로 223-1  20090421         NaN   
86589                     9.2   서울 성동구 용답중앙15길 13.  20131128         NaN   

          행정동 HPSN_MCT_ZCD_NM  폐업여부  n_months  
0      성수1가1동          외식(기타)     0        24  
1      성수1가1동          외식(기타)     0        24  
2         마장동           주점/유흥     0        24  
3         마장동           주점/유흥     0        24  
4         마장동           주점/유흥     0        24  
...       ...             ...   ...       ...  
86585   금호1가동          카페/디저트     0        24  
86586     옥수동              한식     0        24  
86587  성수1가2동              한식     1        12  
86588     사근동          간편식/분식 

In [ ]:
# 불필요한 컬럼 삭제
final_df = final_df.drop(columns=['n_months_x', 'n_months_y'], errors='ignore')

# 결과 확인
print(final_df[['ENCODED_MCT','n_months']].head())

  ENCODED_MCT  n_months
0  000F03E44A        24
1  000F03E44A        24
2  002816BA73        24
3  002816BA73        24
4  002816BA73        24


In [9]:
col_map = {
    'ENCODED_MCT': '가맹점ID',
    'MCT_BSE_AR': '가맹점주소',
    'MCT_NM': '가맹점명',
    'MCT_BRD_NUM': '브랜드코드',
    'MCT_SIGUNGU_NM': '가맹점지역',
    'HPSN_MCT_ZCD_NM': '업종',
    'HPSN_MCT_BZN_CD_NM': '상권',
    'ARE_D': '개설일',
    'MCT_ME_D': '폐업일',
    
    'TA_YM': '기준년월',
    'MCT_OPE_MS_CN': '운영개월수구간',
    'RC_M1_SAA': '매출금액구간',
    'RC_M1_TO_UE_CT': '매출건수구간',
    'RC_M1_UE_CUS_CN': '고객수구간',
    'RC_M1_AV_NP_AT': '객단가구간',
    'APV_CE_RAT': '취소율구간',
    'DLV_SAA_RAT': '배달매출비율',
    
    'M1_SME_RY_SAA_RAT': '동일업종매출비율',
    'M1_SME_RY_CNT_RAT': '동일업종매출건수비율',
    'M12_SME_RY_SAA_PCE_RT': '동일업종매출순위비율',
    'M12_SME_BZN_SAA_PCE_RT': '동일상권매출순위비율',
    'M12_SME_RY_ME_MCT_RAT': '동일업종해지비율',
    'M12_SME_BZN_ME_MCT_RAT': '동일상권해지비율',
    
    '폐업여부': '폐업여부',
    'n_months': '데이터개월수'
}


In [10]:
# 컬럼명 변경
final_df = final_df.rename(columns=col_map)

# 결과 확인
print("컬럼명 변경 완료 ✅")
print(final_df.columns.tolist())


컬럼명 변경 완료 ✅
['가맹점ID', '기준년월', '운영개월수구간', '매출금액구간', '매출건수구간', '동일업종매출비율', '동일업종매출건수비율', '동일업종매출순위비율', '동일상권매출순위비율', '동일업종해지비율', '동일상권해지비율', '가맹점주소', '개설일', '폐업일', '행정동', '업종', '폐업여부', '데이터개월수']


# 개월수 별 모델 비교

데이터는 24개월의 데이터를 주며 다른 기준을 정해주지 않았다. 이에 우리는 통상적 경보를 잘 적용할 수 있는 개월수인 2-5개월의 데이터를 각자 최적의 하이퍼파라미터로 모델링 후 학습-평가를 진행하였을 때 가장 좋은 평가의 개월수 모델을 선정하여 진행하기로 했다.

# 불친절 데이터 정제

대회에서 제공받은 데이터에는 23-24년 사이의 정상가게 데이터(24개월)와 폐업가게 데이터(1~24개월 : 폐업 이후에는 데이터가 없음)의 데이터의 개월수가 불규칙적으로 섞여있다. 
따라서 5개월 기준 적은 데이터는 남은 값들의 평균값으로 대체 후 5개월을 채웠고 5개월을 넘는 데이터는 최근 5개월 이후 데이터는 제외하여 데이터프레임을 만들었다.

In [11]:
import pandas as pd
import numpy as np

# 1️⃣ 수치형 컬럼만 추출
numeric_cols = final_df.select_dtypes(include=[np.number]).columns.tolist()

# 2️⃣ '기준년월'을 정수형으로 변환
final_df['기준년월'] = final_df['기준년월'].astype(int)

# 3️⃣ 점포별 그룹화
grouped = final_df.groupby('가맹점ID')

processed_list = []

for store_id, group in grouped:
    group_sorted = group.sort_values('기준년월', ascending=False).reset_index(drop=True)
    n_months = len(group_sorted)

    # 공통 메타 컬럼 (수치형 제외)
    meta_cols = [col for col in group.columns if col not in numeric_cols]

    # ✅ 5개월 이상 → 최근 5개월만 사용
    if n_months >= 5:
        processed = group_sorted.head(5)

    # ✅ 4개월 → 평균값으로 이전 1개월 생성
    elif n_months == 4:
        processed = group_sorted.copy()

        avg_row = group_sorted[numeric_cols].mean().to_dict()
        new_row = {col: avg_row.get(col, np.nan) for col in numeric_cols}

        # 메타 정보는 최신 행 기준으로 복사
        for col in meta_cols:
            new_row[col] = group_sorted.iloc[0][col]

        # 기준년월은 가장 오래된 달보다 1 작게
        new_row['기준년월'] = group_sorted['기준년월'].min() - 1

        processed = pd.concat([processed, pd.DataFrame([new_row])], ignore_index=True)

    # ✅ 3개월 → 평균값으로 이전 2개월 생성
    elif n_months == 3:
        processed = group_sorted.copy()
        avg_row = group_sorted[numeric_cols].mean().to_dict()
        new_rows = []
        min_month = group_sorted['기준년월'].min()

        for i in range(2):
            new_row = {col: avg_row.get(col, np.nan) for col in numeric_cols}
            for col in meta_cols:
                new_row[col] = group_sorted.iloc[0][col]
            new_row['기준년월'] = min_month - (i + 1)
            new_rows.append(new_row)

        processed = pd.concat([processed, pd.DataFrame(new_rows)], ignore_index=True)

    # ✅ 2개월 → 평균값으로 이전 3개월 생성
    elif n_months == 2:
        processed = group_sorted.copy()
        avg_row = group_sorted[numeric_cols].mean().to_dict()
        new_rows = []
        min_month = group_sorted['기준년월'].min()

        for i in range(3):
            new_row = {col: avg_row.get(col, np.nan) for col in numeric_cols}
            for col in meta_cols:
                new_row[col] = group_sorted.iloc[0][col]
            new_row['기준년월'] = min_month - (i + 1)
            new_rows.append(new_row)

        processed = pd.concat([processed, pd.DataFrame(new_rows)], ignore_index=True)

    # ✅ 1개월 이하 → 제외
    else:
        continue

    processed_list.append(processed)

# 4️⃣ 최종 병합
standardized_df = pd.concat(processed_list, ignore_index=True)

# 5️⃣ 최신순 정렬
standardized_df = standardized_df.sort_values(['가맹점ID', '기준년월'], ascending=[True, False]).reset_index(drop=True)


In [12]:
standardized_df

가맹점ID    기준년월   운영개월수구간              매출금액구간              매출건수구간  \
0      000F03E44A  202412  4_50-75%  6_90%초과(하위 10% 이하)            5_75-90%   
1      000F03E44A  202411  4_50-75%  6_90%초과(하위 10% 이하)            5_75-90%   
2      000F03E44A  202410  4_50-75%            5_75-90%            5_75-90%   
3      000F03E44A  202409  4_50-75%            5_75-90%            5_75-90%   
4      000F03E44A  202408  4_50-75%  6_90%초과(하위 10% 이하)  6_90%초과(하위 10% 이하)   
...           ...     ...       ...                 ...                 ...   
20790  FFF0DAC445  202412  3_25-50%            2_10-25%            3_25-50%   
20791  FFF0DAC445  202411  3_25-50%            2_10-25%            3_25-50%   
20792  FFF0DAC445  202410  3_25-50%            2_10-25%            3_25-50%   
20793  FFF0DAC445  202409  3_25-50%             1_10%이하            3_25-50%   
20794  FFF0DAC445  202408  3_25-50%            2_10-25%            3_25-50%   

       동일업종매출비율  동일업종매출건수비율  동일업종매출순위비율  동일상권매출순위비율  동일업종해지비율  동일상권해지비율  \
0           0.7         4.1        86.8        62.8      17.4       9.0   
1           0.3         1.0        87.0        62.8      17.2       9.1   
2           2.2         3.9        87.1        62.4      17.2       9.2   
3           8.8         4.3        87.5        63.9      17.2       9.3   
4           0.0         0.0        88.8        65.6      17.2       9.1   
...         ...         ...         ...         ...       ...       ...   
20790     319.7       398.3         3.1         7.9      17.0       7.8   
20791     355.0       407.1         2.7         7.3      16.6       8.2   
20792     333.6       386.6         2.7         7.3      16.9       8.1   
20793     424.3       456.2         2.7         7.3      16.9       8.3   
20794     294.8       361.7         2.8         7.6      16.6       8.0   

                     가맹점주소         개설일  폐업일     행정동      업종  폐업여부  데이터개월수  
0      서울특별시 성동구 왕십리로4가길 9  20220225.0  NaN  성수1가1동  외식(기타)   0.0    24.0  
1      서울특별시 성동구 왕십리로4가길 9  20220225.0  NaN  성수1가1동  외식(기타)   0.0    24.0  
2      서울특별시 성동구 왕십리로4가길 9  20220225.0  NaN  성수1가1동  외식(기타)   0.0    24.0  
3      서울특별시 성동구 왕십리로4가길 9  20220225.0  NaN  성수1가1동  외식(기타)   0.0    24.0  
4      서울특별시 성동구 왕십리로4가길 9  20220225.0  NaN  성수1가1동  외식(기타)   0.0    24.0  
...                    ...         ...  ...     ...     ...   ...     ...  
20790  서울 성동구 아차산로7길 10-1.  20160704.0  NaN  성수2가3동  간편식/분식   0.0    24.0  
20791  서울 성동구 아차산로7길 10-1.  20160704.0  NaN  성수2가3동  간편식/분식   0.0    24.0  
20792  서울 성동구 아차산로7길 10-1.  20160704.0  NaN  성수2가3동  간편식/분식   0.0    24.0  
20793  서울 성동구 아차산로7길 10-1.  20160704.0  NaN  성수2가3동  간편식/분식   0.0    24.0  
20794  서울 성동구 아차산로7길 10-1.  20160704.0  NaN  성수2가3동  간편식/분식   0.0    24.0  

[20795 rows x 18 columns]

In [13]:
# 폐업여부 복원: 폐업일이 NaN이 아니면 1, 그렇지 않으면 0
standardized_df["폐업여부"] = standardized_df["폐업일"].notna().astype(int)

standardized_df

가맹점ID    기준년월   운영개월수구간              매출금액구간              매출건수구간  \
0      000F03E44A  202412  4_50-75%  6_90%초과(하위 10% 이하)            5_75-90%   
1      000F03E44A  202411  4_50-75%  6_90%초과(하위 10% 이하)            5_75-90%   
2      000F03E44A  202410  4_50-75%            5_75-90%            5_75-90%   
3      000F03E44A  202409  4_50-75%            5_75-90%            5_75-90%   
4      000F03E44A  202408  4_50-75%  6_90%초과(하위 10% 이하)  6_90%초과(하위 10% 이하)   
...           ...     ...       ...                 ...                 ...   
20790  FFF0DAC445  202412  3_25-50%            2_10-25%            3_25-50%   
20791  FFF0DAC445  202411  3_25-50%            2_10-25%            3_25-50%   
20792  FFF0DAC445  202410  3_25-50%            2_10-25%            3_25-50%   
20793  FFF0DAC445  202409  3_25-50%             1_10%이하            3_25-50%   
20794  FFF0DAC445  202408  3_25-50%            2_10-25%            3_25-50%   

       동일업종매출비율  동일업종매출건수비율  동일업종매출순위비율  동일상권매출순위비율  동일업종해지비율  동일상권해지비율  \
0           0.7         4.1        86.8        62.8      17.4       9.0   
1           0.3         1.0        87.0        62.8      17.2       9.1   
2           2.2         3.9        87.1        62.4      17.2       9.2   
3           8.8         4.3        87.5        63.9      17.2       9.3   
4           0.0         0.0        88.8        65.6      17.2       9.1   
...         ...         ...         ...         ...       ...       ...   
20790     319.7       398.3         3.1         7.9      17.0       7.8   
20791     355.0       407.1         2.7         7.3      16.6       8.2   
20792     333.6       386.6         2.7         7.3      16.9       8.1   
20793     424.3       456.2         2.7         7.3      16.9       8.3   
20794     294.8       361.7         2.8         7.6      16.6       8.0   

                     가맹점주소         개설일  폐업일     행정동      업종  폐업여부  데이터개월수  
0      서울특별시 성동구 왕십리로4가길 9  20220225.0  NaN  성수1가1동  외식(기타)     0    24.0  
1      서울특별시 성동구 왕십리로4가길 9  20220225.0  NaN  성수1가1동  외식(기타)     0    24.0  
2      서울특별시 성동구 왕십리로4가길 9  20220225.0  NaN  성수1가1동  외식(기타)     0    24.0  
3      서울특별시 성동구 왕십리로4가길 9  20220225.0  NaN  성수1가1동  외식(기타)     0    24.0  
4      서울특별시 성동구 왕십리로4가길 9  20220225.0  NaN  성수1가1동  외식(기타)     0    24.0  
...                    ...         ...  ...     ...     ...   ...     ...  
20790  서울 성동구 아차산로7길 10-1.  20160704.0  NaN  성수2가3동  간편식/분식     0    24.0  
20791  서울 성동구 아차산로7길 10-1.  20160704.0  NaN  성수2가3동  간편식/분식     0    24.0  
20792  서울 성동구 아차산로7길 10-1.  20160704.0  NaN  성수2가3동  간편식/분식     0    24.0  
20793  서울 성동구 아차산로7길 10-1.  20160704.0  NaN  성수2가3동  간편식/분식     0    24.0  
20794  서울 성동구 아차산로7길 10-1.  20160704.0  NaN  성수2가3동  간편식/분식     0    24.0  

[20795 rows x 18 columns]

# 모델 복잡도 제어

데이터의 n수가 적기 때문에 변수를 늘려 복잡도를 제어했다.

In [14]:
# 기준 컬럼명 지정
target_col = '동일업종매출비율'  # 기준 지표
change_col = '매출변화율'       # 새로 만들 변화율 컬럼명

# 결과 저장용 리스트
change_rows = []

# 점포별 처리
for store_id, group in standardized_df.groupby('가맹점ID'):
    # 최근 기준년월 순으로 정렬
    group_sorted = group.sort_values('기준년월', ascending=False).reset_index(drop=True)

    # ✅ 5개월 데이터 존재하는 경우만 처리
    if len(group_sorted) == 5:
        try:
            v0 = group_sorted.loc[0, target_col]  # 최근달
            v1 = group_sorted.loc[1, target_col]
            v2 = group_sorted.loc[2, target_col]
            v3 = group_sorted.loc[3, target_col]
            v4 = group_sorted.loc[4, target_col]  # 가장 오래된 달

            # 변화율 계산
            r1 = (v0 - v1) / v1 if v1 != 0 else 0
            r2 = (v1 - v2) / v2 if v2 != 0 else 0
            r3 = (v2 - v3) / v3 if v3 != 0 else 0
            r4 = (v3 - v4) / v4 if v4 != 0 else 0

            # 평균 변화율 계산
            avg_change = (r1 + r2 + r3 + r4) / 4

            # 전체 그룹에 동일한 변화율 추가
            group_sorted[change_col] = avg_change

        except Exception:
            group_sorted[change_col] = 0  # 예외 시 0 처리

        change_rows.append(group_sorted)

# ✅ 결과 병합
final_with_change = pd.concat(change_rows, ignore_index=True)


In [15]:
import pandas as pd
import numpy as np

def get_large_category(sector):
    """
    세부 업종(HPSN_MCT_ZCD_NM) 문자열을 입력받아
    대분류 문자열을 반환합니다.
    """
    if pd.isna(sector):
        return np.nan 
    
    # 1. 한식
    if '한식-' in sector or sector in ['백반/가정식', '한정식', '기사식당', '한식뷔페', '구내식당/푸드코트']:
        return '한식'
    
    # 2. 카페/디저트
    if sector in ['카페', '커피전문점', '테이크아웃커피', '테마카페', '차', '주스',
                  '베이커리', '아이스크림/빙수', '도너츠', '떡/한과', '마카롱', '와플/크로플', '탕후루']:
        return '카페/디저트'
    
    # 3. 외식 (기타)
    if '일식-' in sector or sector in ['일식당', '양식', '중식당', '동남아/인도음식', '기타세계요리', '스테이크', 
                                   '중식-딤섬/중식만두', '중식-훠궈/마라탕']:
        return '외식(기타)'
    
    # 4. 간편식/분식
    if sector in ['분식', '치킨', '피자', '햄버거', '샌드위치/토스트', '도시락', '꼬치구이']:
        return '간편식/분식'
        
    # 5. 주점/유흥
    if sector in ['호프/맥주', '요리주점', '이자카야', '민속주점', '포장마차', '와인바', 
                  '일반 유흥주점', '룸살롱/단란주점']:
        return '주점/유흥'

    # 6. 소매/식재료
    if sector in ['축산물', '식료품', '농산물', '수산물', '청과물', '미곡상', '건어물', '유제품', '반찬']:
        return '소매/식재료'

    # 7. 기타/제조
    if sector in ['건강식품', '건강원', '인삼제품', '식품 제조', '떡/한과 제조', '담배', '와인샵', '주류']:
        return '기타/제조'

    return '기타(미분류)'

# --- 메인 코드 실행 ---

input_file = 'big_data_with_hdong_final.csv'
output_file = 'big_data_with_large_categories_FINAL.csv'

print("--- 스크립트 실행 시작 ---")

try:
    print(f"'{input_file}' 파일을 읽고 있습니다...")
    df = pd.read_csv(input_file)
    print("파일 읽기 완료.")

    if 'HPSN_MCT_ZCD_NM' not in df.columns:
        # 'HPSN_MCT_ZCD_NM' 컬럼이 없으면 여기서 KeyError를 발생시킴
        raise KeyError
    
    print(f"'HPSN_MCT_ZCD_NM' 컬럼을 대분류로 변환합니다...")
    df['HPSN_MCT_ZCD_NM'] = df['HPSN_MCT_ZCD_NM'].map(get_large_category)
    print("변환 완료.")

    df.to_csv(output_file, index=False, encoding='utf-8-sig')
    
    print(f"\n--- 작업 성공 ---")
    print(f"'{output_file}' 이름으로 파일 저장을 완료했습니다.")
    print("파일 내용을 확인해보세요.")

except FileNotFoundError:
    print(f"\n[오류] '{input_file}' 파일을 찾을 수 없습니다.")
    print("이 스크립트 파일과 원본 파일이 같은 폴더에 있는지 확인해주세요.")
    
except KeyError:
    # [오류 수정] as e: 부분을 제거하고 가장 단순한 구문으로 변경했습니다.
    print(f"\n[오류] 'HPSN_MCT_ZCD_NM' 컬럼을 찾지 못했습니다.")
    print("원본 파일에 해당 컬럼명이 정확히 존재하는지 확인해주세요.")
    
except Exception as e:
    print(f"\n[알 수 없는 오류] 작업 중 오류가 발생했습니다: {e}")

print("--- 스크립트 실행 종료 ---")

--- 스크립트 실행 시작 ---
'big_data_with_hdong_final.csv' 파일을 읽고 있습니다...
파일 읽기 완료.
'HPSN_MCT_ZCD_NM' 컬럼을 대분류로 변환합니다...
변환 완료.

--- 작업 성공 ---
'big_data_with_large_categories_FINAL.csv' 이름으로 파일 저장을 완료했습니다.
파일 내용을 확인해보세요.
--- 스크립트 실행 종료 ---


In [16]:
final_with_change['행정동']

0        성수1가1동
1        성수1가1동
2        성수1가1동
3        성수1가1동
4        성수1가1동
          ...  
20790    성수2가3동
20791    성수2가3동
20792    성수2가3동
20793    성수2가3동
20794    성수2가3동
Name: 행정동, Length: 20795, dtype: object

In [17]:
import pandas as pd

# 이미 전처리된 standardized_df 사용
df = final_with_change.copy()

# 1️⃣ 필요한 컬럼만 추출
cols = [
    '가맹점ID', '기준년월', '가맹점주소', '업종', '폐업여부',
    '동일업종매출비율', '동일업종매출순위비율',
    '동일상권매출순위비율', '동일업종해지비율',
    '동일상권해지비율', '데이터개월수','행정동'
    
]
df = df[cols]

# 3️⃣ 같은 행정동 + 같은 업종 단위로 그룹화
grouped = (
    df.groupby(['행정동', '업종'])
    .agg({
        '가맹점ID': 'nunique',
        '폐업여부': 'mean',
        '동일업종매출비율': 'mean',
        '동일상권매출순위비율': 'mean',
        '동일업종해지비율': 'mean'
    })
    .reset_index()
)

# 4️⃣ 컬럼명 보기 좋게 변경
grouped = grouped.rename(columns={
    '가맹점ID': '가맹점수',
    '폐업여부': '평균폐업률',
    '동일업종매출비율': '평균동일업종매출비율',
    '동일상권매출순위비율': '평균상권매출순위비율',
    '동일업종해지비율': '평균업종해지비율'
})

# 5️⃣ 결과 확인
print(grouped.head(10))

       행정동      업종  가맹점수     평균폐업률  평균동일업종매출비율  평균상권매출순위비율   평균업종해지비율
0     14-2      한식     1  0.000000  408.280000    2.840000  18.320000
1     17-7  카페/디저트     1  0.000000  714.360000   11.820000  16.380000
2      178  외식(기타)     1  0.000000   64.260000   44.840000  18.160000
3    19-54      한식     1  0.000000   71.260000   47.920000  18.320000
4      207      한식     1  0.000000  135.500000   12.680000  16.880000
5  289-175  외식(기타)     1  0.000000  204.300000   10.740000  17.240000
6    328-7  카페/디저트     1  0.000000   81.920000   31.480000  14.920000
7     47-1      한식     1  0.000000  496.160000    4.060000  16.880000
8      5-3  외식(기타)     1  0.000000  267.680000    4.040000  18.160000
9    금호1가동  간편식/분식    11  0.090909   83.718182   22.896364  18.005455


In [18]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.cluster import KMeans
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score, f1_score
import lightgbm as lgb

dataset3 = pd.read_csv('big_data_set3_f.csv')

# =========================================
# 2️⃣ 결측치 처리 및 이상값 정리
# =========================================
final_with_change.replace(['-', ' ', ''], np.nan, inplace=True)
final_with_change = final_with_change.replace(-999999.9, np.nan)

dataset3 = dataset3.replace(-999999.9, np.nan)

# =========================================
# 3️⃣ 행정동별 군집화
# =========================================
# 군집화에 사용할 주요 지표 선정 (필요시 조정)
cluster_features = [
    '매출변화율',
    '동일업종매출비율',
    '동일상권매출순위비율',
    '동일상권해지비율'
]

# 행정동별 평균 계산
grouped = (
    final_with_change
    .groupby('행정동')[cluster_features]
    .mean()
    .reset_index()
)

# 표준화 전에 NaN을 평균값으로 채우기
grouped[cluster_features] = grouped[cluster_features].fillna(grouped[cluster_features].mean())

# 표준화
scaler = StandardScaler()
scaled = scaler.fit_transform(grouped[cluster_features])

# 군집화 수행
kmeans = KMeans(n_clusters=5, random_state=42)
grouped['cluster'] = kmeans.fit_predict(scaled)

# 군집 결과 병합
final_with_change = final_with_change.merge(grouped[['행정동', 'cluster']], on='행정동', how='left')

C:\Users\hanse\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [19]:
# dataset3 컬럼명 변경
dataset3 = dataset3.rename(columns={
    'ENCODED_MCT': '가맹점ID',
    'TA_YM': '기준년월'
})

# 변경 확인
print(dataset3.columns.tolist())

['가맹점ID', '기준년월', 'M12_MAL_1020_RAT', 'M12_MAL_30_RAT', 'M12_MAL_40_RAT', 'M12_MAL_50_RAT', 'M12_MAL_60_RAT', 'M12_FME_1020_RAT', 'M12_FME_30_RAT', 'M12_FME_40_RAT', 'M12_FME_50_RAT', 'M12_FME_60_RAT', 'MCT_UE_CLN_REU_RAT', 'MCT_UE_CLN_NEW_RAT', 'RC_M1_SHC_RSD_UE_CLN_RAT', 'RC_M1_SHC_WP_UE_CLN_RAT', 'RC_M1_SHC_FLP_UE_CLN_RAT']


In [20]:
# =========================================
# 4️⃣ 고객 데이터 병합
# =========================================
merged_df = pd.merge(final_with_change, dataset3, on=['가맹점ID', '기준년월'], how='left')

# =========================================
# 5️⃣ Label Encoding (문자형 변수)
# =========================================
cat_cols = merged_df.select_dtypes(include='object').columns
encoder = LabelEncoder()
for col in cat_cols:
    merged_df[col] = encoder.fit_transform(merged_df[col].astype(str))

# =========================================
# 6️⃣ 불필요한 변수 제거
# =========================================
drop_cols = ['가맹점주소', '데이터부족', '개설일', '폐업일']
merged_df = merged_df.drop(columns=[c for c in drop_cols if c in merged_df.columns])

# =========================================
# 7️⃣ 결측치 처리
# =========================================
merged_df = merged_df.fillna(merged_df.median(numeric_only=True))

# =========================================
# ✅ 최종 확인
# =========================================
print("최종 shape:", merged_df.shape)
print("결측치 비율:", merged_df.isna().mean().mean())
print("컬럼 수:", len(merged_df.columns))
print(merged_df.head(3))

최종 shape: (20795, 32)
결측치 비율: 0.0
컬럼 수: 32
   가맹점ID    기준년월  운영개월수구간  매출금액구간  매출건수구간  동일업종매출비율  동일업종매출건수비율  동일업종매출순위비율  \
0      0  202412        3       5       4       0.7         4.1        86.8   
1      0  202411        3       5       4       0.3         1.0        87.0   
2      0  202410        3       4       4       2.2         3.9        87.1   

   동일상권매출순위비율  동일업종해지비율  ...  M12_FME_1020_RAT  M12_FME_30_RAT  \
0        62.8      17.4  ...           11.1250          8.9000   
1        62.8      17.2  ...            9.4794          8.2838   
2        62.4      17.2  ...           10.9000          8.4584   

   M12_FME_40_RAT  M12_FME_50_RAT  M12_FME_60_RAT  MCT_UE_CLN_REU_RAT  \
0          4.4500          11.125          8.9000               10.87   
1          4.7397          10.675          9.4794                9.30   
2          4.8396          10.900          8.4584                9.30   

   MCT_UE_CLN_NEW_RAT  RC_M1_SHC_RSD_UE_CLN_RAT  RC_M1_SHC_WP_UE_CLN_RAT  \
0     

In [21]:
# dataset3 컬럼명 전체 한글화
dataset3 = dataset3.rename(columns={
    'M12_MAL_1020_RAT': '남성 20대이하 고객 비중',
    'M12_MAL_30_RAT': '남성 30대 고객 비중',
    'M12_MAL_40_RAT': '남성 40대 고객 비중',
    'M12_MAL_50_RAT': '남성 50대 고객 비중',
    'M12_MAL_60_RAT': '남성 60대이상 고객 비중',
    'M12_FME_1020_RAT': '여성 20대이하 고객 비중',
    'M12_FME_30_RAT': '여성 30대 고객 비중',
    'M12_FME_40_RAT': '여성 40대 고객 비중',
    'M12_FME_50_RAT': '여성 50대 고객 비중',
    'M12_FME_60_RAT': '여성 60대이상 고객 비중',
    'MCT_UE_CLN_REU_RAT': '재방문 고객 비중',
    'MCT_UE_CLN_NEW_RAT': '신규 고객 비중',
    'RC_M1_SHC_RSD_UE_CLN_RAT': '거주 이용 고객 비율',
    'RC_M1_SHC_WP_UE_CLN_RAT': '직장 이용 고객 비율',
    'RC_M1_SHC_FLP_UE_CLN_RAT': '유동인구 이용 고객 비율'
})

# 변경 결과 확인
print("변경된 컬럼명:")
print(merged_df.columns.tolist())


변경된 컬럼명:
['가맹점ID', '기준년월', '운영개월수구간', '매출금액구간', '매출건수구간', '동일업종매출비율', '동일업종매출건수비율', '동일업종매출순위비율', '동일상권매출순위비율', '동일업종해지비율', '동일상권해지비율', '행정동', '업종', '폐업여부', '데이터개월수', '매출변화율', 'cluster', 'M12_MAL_1020_RAT', 'M12_MAL_30_RAT', 'M12_MAL_40_RAT', 'M12_MAL_50_RAT', 'M12_MAL_60_RAT', 'M12_FME_1020_RAT', 'M12_FME_30_RAT', 'M12_FME_40_RAT', 'M12_FME_50_RAT', 'M12_FME_60_RAT', 'MCT_UE_CLN_REU_RAT', 'MCT_UE_CLN_NEW_RAT', 'RC_M1_SHC_RSD_UE_CLN_RAT', 'RC_M1_SHC_WP_UE_CLN_RAT', 'RC_M1_SHC_FLP_UE_CLN_RAT']


In [22]:
merged_df

가맹점ID    기준년월  운영개월수구간  매출금액구간  매출건수구간  동일업종매출비율  동일업종매출건수비율  \
0          0  202412        3       5       4       0.7         4.1   
1          0  202411        3       5       4       0.3         1.0   
2          0  202410        3       4       4       2.2         3.9   
3          0  202409        3       4       4       8.8         4.3   
4          0  202408        3       5       5       0.0         0.0   
...      ...     ...      ...     ...     ...       ...         ...   
20790   4158  202412        2       1       2     319.7       398.3   
20791   4158  202411        2       1       2     355.0       407.1   
20792   4158  202410        2       1       2     333.6       386.6   
20793   4158  202409        2       0       2     424.3       456.2   
20794   4158  202408        2       1       2     294.8       361.7   

       동일업종매출순위비율  동일상권매출순위비율  동일업종해지비율  ...  M12_FME_1020_RAT  \
0            86.8        62.8      17.4  ...           11.1250   
1            87.0        62.8      17.2  ...            9.4794   
2            87.1        62.4      17.2  ...           10.9000   
3            87.5        63.9      17.2  ...           10.9626   
4            88.8        65.6      17.2  ...           13.3750   
...           ...         ...       ...  ...               ...   
20790         3.1         7.9      17.0  ...           17.8407   
20791         2.7         7.3      16.6  ...           17.9536   
20792         2.7         7.3      16.9  ...           18.8201   
20793         2.7         7.3      16.9  ...           18.8600   
20794         2.8         7.6      16.6  ...           19.2045   

       M12_FME_30_RAT  M12_FME_40_RAT  M12_FME_50_RAT  M12_FME_60_RAT  \
0              8.9000          4.4500         11.1250          8.9000   
1              8.2838          4.7397         10.6750          9.4794   
2              8.4584          4.8396         10.9000          8.4584   
3              8.2446          5.4813         12.3669          8.2446   
4              8.3460          6.6875         15.0335         10.0580   
...               ...             ...             ...             ...   
20790         18.1634          5.3476          3.4575          1.2908   
20791         17.9536          5.3128          3.4350          1.1450   
20792         18.1196          5.3238          3.3624          1.0741   
20793         17.4340          5.3820          3.3580          1.0120   
20794         17.4375          5.5800          3.2550          1.0695   

       MCT_UE_CLN_REU_RAT  MCT_UE_CLN_NEW_RAT  RC_M1_SHC_RSD_UE_CLN_RAT  \
0                   10.87                6.52                       0.0   
1                    9.30                2.33                       0.0   
2                    9.30                9.30                      66.7   
3                   10.26                7.69                       0.0   
4                    8.11                0.00                      32.1   
...                   ...                 ...                       ...   
20790               10.41                7.68                      24.0   
20791               10.28                7.96                      17.9   
20792               10.27                7.48                      16.8   
20793                9.62                8.07                      21.5   
20794                9.60                6.24                      16.0   

       RC_M1_SHC_WP_UE_CLN_RAT  RC_M1_SHC_FLP_UE_CLN_RAT  
0                         25.0                      75.0  
1                          0.0                     100.0  
2                          0.0                      33.3  
3                         50.0                      50.0  
4                          8.2                      50.0  
...                        ...                       ...  
20790                     15.7                      60.3  
20791                     19.5                      62.6  
20792                     16.0                      

In [23]:
df = merged_df

In [24]:
df['동일동내_매출순위'] = df.groupby('행정동')['매출변화율'].rank(ascending=False, method='dense')

# 각 행정동 내 전체 점포 수
df['동별_점포수'] = df.groupby('행정동')['매출변화율'].transform('count')

# 순위를 점포 수로 나눈 비율 (0~1 사이)
df['동일동매출순위비율'] = df['동일동내_매출순위'] / df['동별_점포수']

# 결과 확인
df[['행정동', '매출변화율', '동일동내_매출순위', '동별_점포수', '동일동매출순위비율']].head(10)

행정동     매출변화율  동일동내_매출순위  동별_점포수  동일동매출순위비율
0   21 -0.070076      146.0     880   0.165909
1   21 -0.070076      146.0     880   0.165909
2   21 -0.070076      146.0     880   0.165909
3   21 -0.070076      146.0     880   0.165909
4   21 -0.070076      146.0     880   0.165909
5   18  0.092608      200.0    2090   0.095694
6   18  0.092608      200.0    2090   0.095694
7   18  0.092608      200.0    2090   0.095694
8   18  0.092608      200.0    2090   0.095694
9   18  0.092608      200.0    2090   0.095694

In [25]:
type(df)

pandas.core.frame.DataFrame

In [26]:
# 영어 변수명 → 한글 변수명 매핑 딕셔너리
col_map = {
    'M12_MAL_1020_RAT': '남성 20대이하 고객 비중',
    'M12_MAL_30_RAT': '남성 30대 고객 비중',
    'M12_MAL_40_RAT': '남성 40대 고객 비중',
    'M12_MAL_50_RAT': '남성 50대 고객 비중',
    'M12_MAL_60_RAT': '남성 60대이상 고객 비중',
    'M12_FME_1020_RAT': '여성 20대이하 고객 비중',
    'M12_FME_30_RAT': '여성 30대 고객 비중',
    'M12_FME_40_RAT': '여성 40대 고객 비중',
    'M12_FME_50_RAT': '여성 50대 고객 비중',
    'M12_FME_60_RAT': '여성 60대이상 고객 비중',
    'MCT_UE_CLN_REU_RAT': '재방문 고객 비중',
    'MCT_UE_CLN_NEW_RAT': '신규 고객 비중',
    'RC_M1_SHC_RSD_UE_CLN_RAT': '거주 이용 고객 비율',
    'RC_M1_SHC_WP_UE_CLN_RAT': '직장 이용 고객 비율',
    'RC_M1_SHC_FLP_UE_CLN_RAT': '유동인구 이용 고객 비율'
}

# ✅ 컬럼명 변경
df.rename(columns=col_map, inplace=True)

# ✅ 변경 결과 확인
print(df.columns)


Index(['가맹점ID', '기준년월', '운영개월수구간', '매출금액구간', '매출건수구간', '동일업종매출비율',
       '동일업종매출건수비율', '동일업종매출순위비율', '동일상권매출순위비율', '동일업종해지비율', '동일상권해지비율', '행정동',
       '업종', '폐업여부', '데이터개월수', '매출변화율', 'cluster', '남성 20대이하 고객 비중',
       '남성 30대 고객 비중', '남성 40대 고객 비중', '남성 50대 고객 비중', '남성 60대이상 고객 비중',
       '여성 20대이하 고객 비중', '여성 30대 고객 비중', '여성 40대 고객 비중', '여성 50대 고객 비중',
       '여성 60대이상 고객 비중', '재방문 고객 비중', '신규 고객 비중', '거주 이용 고객 비율', '직장 이용 고객 비율',
       '유동인구 이용 고객 비율', '동일동내_매출순위', '동별_점포수', '동일동매출순위비율'],
      dtype='object')


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
import pandas as pd

# 1️⃣ 사용할 데이터 지정 (불필요한 키 변수 제외)
X = df.drop(columns=['가맹점ID', '기준년월', '폐업여부'])
y = df['폐업여부']


In [28]:
# 2️⃣ 범주형 변수 인코딩
cat_cols = X.select_dtypes(include='object').columns
for col in cat_cols:
    le = LabelEncoder()
    X[col] = le.fit_transform(X[col].astype(str))

# 3️⃣ 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [29]:
# train/test 나누기 전에 원본 데이터에서 분리 기준 확인
train = df[df['기준년월'] < 202312]
test = df[df['기준년월'] == 202312]

# 중복된 가맹점 확인
overlap = len(set(train['가맹점ID']) & set(test['가맹점ID']))
print(f"중복된 가맹점 수: {overlap}")

중복된 가맹점 수: 2


In [30]:
# train/test 분리
train = df[df['기준년월'] < 202312]
test = df[df['기준년월'] == 202312]

# 중복된 가맹점ID 확인
dup_ids = set(train['가맹점ID']) & set(test['가맹점ID'])
print(f"중복된 가맹점 수: {len(dup_ids)}")

# ✅ test에서 중복 가맹점 제거
test = test[~test['가맹점ID'].isin(dup_ids)]

print(f"제거 후 테스트 가맹점 수: {test['가맹점ID'].nunique()}")


중복된 가맹점 수: 2
제거 후 테스트 가맹점 수: 2


# 모델 선정 이유

CatBoost 모델 선정 이유는 범주형 변수 자동 처리, 빠른 학습, 과적합 방지 등의 이유와 무엇보다 **추가 설정으로 불균형 데이터를 보안할 수 있다는 점**이 큰 장점이였다.
위 데이터는 위기 점포의 비율이 정상 점포 4185개의 데이터 중 127개로 약 3% 수준이다. 즉 심각한 불균형 데이터였다. 따라서 데이터의 불균형을 맞출 모델이 필요했고 CatBoost를 선정했다.

In [31]:
!pip install catboost

In [32]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression

df = df.sort_values(['가맹점ID', '기준년월']).reset_index(drop=True)

num_cols = [
    '동일업종매출비율', '동일업종매출건수비율', '동일업종매출순위비율', '동일상권매출순위비율',
    '동일업종해지비율', '동일상권해지비율', '매출변화율',
    '남성 20대이하 고객 비중', '남성 30대 고객 비중', '남성 40대 고객 비중',
    '남성 50대 고객 비중', '남성 60대이상 고객 비중',
    '여성 20대이하 고객 비중', '여성 30대 고객 비중', '여성 40대 고객 비중',
    '여성 50대 고객 비중', '여성 60대이상 고객 비중',
    '재방문 고객 비중', '신규 고객 비중',
    '거주 이용 고객 비율', '직장 이용 고객 비율', '유동인구 이용 고객 비율',
    '동별_점포수'
]

# -----------------------------
# 변화율, 평균, 표준편차 계산 (index 정렬 안전하게)
# -----------------------------
for col in num_cols:
    grp = df.groupby('가맹점ID')[col]
    df[f'{col}_5개월변화율'] = grp.transform(lambda x: x.pct_change(periods=5))
    df[f'{col}_5개월평균'] = grp.transform(lambda x: x.rolling(window=5, min_periods=1).mean())
    df[f'{col}_5개월표준편차'] = grp.transform(lambda x: x.rolling(window=5, min_periods=1).std())

# -----------------------------
# 기울기(추세) 계산 함수
# -----------------------------
def slope(series):
    """5개월 구간 선형회귀 기울기"""
    if series.isnull().any() or len(series) < 5:
        return np.nan
    x = np.arange(len(series)).reshape(-1, 1)
    y = series.values
    model = LinearRegression().fit(x, y)
    return model.coef_[0]

for col in num_cols:
    df[f'{col}_5개월추세'] = (
        df.groupby('가맹점ID')[col]
        .transform(lambda x: x.rolling(window=5, min_periods=5).apply(slope, raw=False))
    )

# -----------------------------
# 결측/무한값 정리
# -----------------------------
df.replace([np.inf, -np.inf], np.nan, inplace=True)

print(f"✅ 생성된 변화율/추세 피처 수: {len(num_cols) * 4}")
print(f"✅ 전체 컬럼 수: {df.shape[1]}")


✅ 생성된 변화율/추세 피처 수: 92
✅ 전체 컬럼 수: 127


In [33]:
# ==========================================
# 상관관계 0.9 이상 변수 제거 (단, '폐업여부'는 보호)
# ==========================================

# 1️⃣ 목적변수 지정
target_col = '폐업여부'

# 2️⃣ 수치형 변수 목록에서 목적변수 제외
num_cols_wo_target = [col for col in num_cols if col != target_col]

# 3️⃣ 상관계수 계산 (절대값 기준)
corr = df[num_cols_wo_target].corr().abs()

# 4️⃣ 상삼각 행렬만 유지
upper = corr.where(np.triu(np.ones(corr.shape), k=1).astype(bool))

# 5️⃣ 상관 0.9 이상인 컬럼 탐색
drop_cols = [col for col in upper.columns if any(upper[col] > 0.9)]

# 6️⃣ 실제 컬럼 삭제 (단, 목적변수는 삭제 대상에서 제외)
df = df.drop(columns=[c for c in drop_cols if c != target_col], errors='ignore')

print(f"📉 상관 0.9 이상 변수 제거 후 남은 컬럼 수: {df.shape[1]}")


📉 상관 0.9 이상 변수 제거 후 남은 컬럼 수: 127


In [34]:
# ==========================================
# 결측률 및 변동성 기준 변수 제거 (단, '폐업여부'는 유지)
# ==========================================

target_col = '폐업여부'

# ✅ 실제 df에 존재하는 컬럼만 유지
num_cols = [c for c in num_cols if c in df.columns and c != target_col]

# ✅ 분산 계산 (너무 낮은 경우 제거 대상)
low_var = df[num_cols].std() < 1e-6

# ✅ 결측률 계산
null_rate = df.isnull().mean()

# ✅ 제거 대상 컬럼 (단, 목적변수는 제외)
drop_feats = null_rate[null_rate > 0.5].index.union(low_var[low_var].index)
drop_feats = [col for col in drop_feats if col != target_col]

# ✅ 실제 제거
df = df.drop(columns=drop_feats, errors='ignore')

print(f"📉 결측률/변동성 낮은 변수 제거 후 컬럼 수: {df.shape[1]}")


📉 결측률/변동성 낮은 변수 제거 후 컬럼 수: 81


In [36]:
df[df['폐업여부'] == 1]

가맹점ID    기준년월  운영개월수구간  매출금액구간  매출건수구간    동일업종매출비율  동일업종매출건수비율  \
380       76  202408        2       4       4   15.300000   17.100000   
381       76  202409        2       4       3   23.300000   19.800000   
382       76  202410        2       3       3   27.600000   24.800000   
383       76  202411        2       3       3   29.600000   29.100000   
384       76  202412        2       4       4    9.100000   17.500000   
...      ...     ...      ...     ...     ...         ...         ...   
20625   4125  202299        5       3       2  101.333333  192.166667   
20626   4125  202300        5       3       2  101.333333  192.166667   
20627   4125  202301        3       2       2  106.500000  201.400000   
20628   4125  202302        3       2       1  121.900000  226.300000   
20629   4125  202303        5       3       2   75.600000  148.800000   

       동일업종매출순위비율  동일상권매출순위비율  동일업종해지비율  ...  신규 고객 비중_5개월평균  \
380     49.600000    0.000000      17.1  ...        2.670000   
381     50.100000    0.000000      17.1  ...        3.765000   
382     50.800000    0.000000      17.1  ...        4.943333   
383     50.800000    0.000000      16.9  ...        5.715000   
384     57.900000    0.000000      16.2  ...        5.740000   
...           ...         ...       ...  ...             ...   
20625   15.066667   24.333333      15.4  ...        6.670000   
20626   15.066667   24.333333      15.4  ...        6.670000   
20627   14.900000   24.400000      14.8  ...        5.970000   
20628   14.900000   24.100000      15.5  ...        6.027500   
20629   15.400000   24.500000      15.9  ...        5.784000   

       신규 고객 비중_5개월표준편차  거주 이용 고객 비율_5개월평균  거주 이용 고객 비율_5개월표준편차  \
380                 NaN          83.300000                  NaN   
381            1.548564          80.100000             4.525483   
382            2.316125          74.966667             9.449515   
383            2.440936          69.975000            12.617283   
384            2.114651          67.640000            12.110244   
...                 ...                ...                  ...   
20625               NaN          32.100000                  NaN   
20626          0.000000          32.100000             0.000000   
20627          1.212436          43.333333            19.456704   
20628          0.996607          48.275000            18.709779   
20629          1.020480          52.020000            18.239161   

       직장 이용 고객 비율_5개월평균  직장 이용 고객 비율_5개월표준편차  유동인구 이용 고객 비율_5개월평균  \
380             0.000000                  NaN               16.700   
381             3.850000             5.444722               16.050   
382             4.533333             4.027820               20.500   
383             4.650000             3.296968               25.375   
384             5.380000             3.288921               26.960   
...                  ...                  ...                  ...   
20625           8.200000                  NaN               50.000   
20626           8.200000             0.000000               50.000   
20627           7.800000             0.692820               42.400   
20628           7.400000             0.979796               39.500   
20629           7.380000             0.849706               36.740   

       유동인구 이용 고객 비율_5개월표준편차  동별_점포수_5개월평균  동별_점포수_5개월표준편차  
380                      NaN         620.0             NaN  
381                 0.919239         620.0             0.0  
382                 7.734985         620.0             0.0  
383                11.616762         620.0             0.0  
384                10.666443         620.0             0.0  
...                      ...           ...             ...  
20625                    NaN        2250.0             NaN  
20626               0.000000        2250.0             0.0  
20627              13.163586        2250.0             0.0  
20628              12.213108        2250.0             0.0  
20629              12.245734        2250.0  

In [38]:
from catboost import CatBoostClassifier, Pool

# 임시 학습 (간단히 돌려서 중요도 확인)
model = CatBoostClassifier(iterations=200, depth=6, learning_rate=0.05, verbose=False)
X = df.drop(columns=['가맹점ID', '기준년월', '폐업여부'])
y = df['폐업여부']

model.fit(X, y)
feat_imp = pd.DataFrame({'feature': X.columns, 'importance': model.feature_importances_})
top_feats = feat_imp.sort_values('importance', ascending=False).head(30)
print(top_feats)

                 feature  importance
29             동일동내_매출순위    6.204637
0                운영개월수구간    5.899979
10                    업종    5.568204
11                데이터개월수    4.877510
58    여성 30대 고객 비중_5개월평균    3.851083
38      동일상권매출순위비율_5개월평균    3.417265
31             동일동매출순위비율    3.166071
62    여성 50대 고객 비중_5개월평균    2.907846
50    남성 40대 고객 비중_5개월평균    2.823974
66       재방문 고객 비중_5개월평균    2.664141
34      동일업종매출건수비율_5개월평균    2.631074
48    남성 30대 고객 비중_5개월평균    2.576391
12                 매출변화율    2.567984
44           매출변화율_5개월평균    2.420767
40        동일업종해지비율_5개월평균    2.399385
68        신규 고객 비중_5개월평균    2.187533
56  여성 20대이하 고객 비중_5개월평균    1.934079
52    남성 50대 고객 비중_5개월평균    1.912373
64  여성 60대이상 고객 비중_5개월평균    1.816218
54  남성 60대이상 고객 비중_5개월평균    1.666876
24             재방문 고객 비중    1.643349
23        여성 60대이상 고객 비중    1.634767
36      동일업종매출순위비율_5개월평균    1.607475
20          여성 30대 고객 비중    1.599258
5             동일업종매출순위비율    1.591479
60    여성 40대 고객 비중_5개월평균    1.590809
6

# 시계열 데이터 처리

시계열 데이터인 위 데이터의 경우 기존의 CatBoost를 시차(lag)·rolling feature 등과 같이 쓰거나 CNN/LSTM으로 빠지는 방법이 있지만 가게별 5개월 데이터를 그룹화하여 하나로 취급 후 샘플링.

In [40]:
# ==========================================
# 📦 라이브러리
# ==========================================
import pandas as pd
import numpy as np
from sklearn.model_selection import GroupKFold
from sklearn.utils import resample
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from catboost import CatBoostClassifier

# ==========================================
# 1️⃣ 언더샘플링 (폐업:비폐업 균형 조정)
# ==========================================
df_minority = df[df['폐업여부'] == 1]
df_majority = df[df['폐업여부'] == 0]

df_majority_down = resample(df_majority,
                            replace=False,
                            n_samples=len(df_minority) * 5,
                            random_state=42)

df_balanced = pd.concat([df_majority_down, df_minority]).sample(frac=1, random_state=42).reset_index(drop=True)

print("언더샘플링 후 데이터 크기:", df_balanced.shape)
print(df_balanced['폐업여부'].value_counts())

# ==========================================
# 2️⃣ 데이터 분리
# ==========================================
X = df_balanced.drop(columns=['가맹점ID', '기준년월', '폐업여부'])
y = df_balanced['폐업여부']
groups = df_balanced['가맹점ID']  # GroupKFold에서 동일 점포 분리 기준

# ==========================================
# 3️⃣ Group K-Fold 설정
# ==========================================
gkf = GroupKFold(n_splits=5)
results = []

# ==========================================
# 4️⃣ 교차검증 루프
# ==========================================
for fold, (train_idx, val_idx) in enumerate(gkf.split(X, y, groups)):
    print(f"\n===== Fold {fold+1} =====")

    X_train, X_val = X.iloc[train_idx], X.iloc[val_idx]
    y_train, y_val = y.iloc[train_idx], y.iloc[val_idx]

    # ------------------------------
    # (A) CatBoost 1단계 학습
    # ------------------------------
    model = CatBoostClassifier(
        iterations=1000,
        depth=8,
        learning_rate=0.01,
        l2_leaf_reg=2,
        subsample=0.9,
        rsm=0.9,
        class_weights=[1, 20],  # 폐업 가중치 ↑
        eval_metric='Recall',
        random_seed=42,
        verbose=False
    )
    model.fit(X_train, y_train)

    # ------------------------------
    # (B) 중요 피처 80개만 선별
    # ------------------------------
    fi = pd.DataFrame({
        'feature': X.columns,
        'importance': model.get_feature_importance()
    })
    top_features = fi.sort_values('importance', ascending=False).head(80)['feature'].tolist()

    X_train_sel = X_train[top_features]
    X_val_sel = X_val[top_features]

    # ------------------------------
    # (C) 상위 피처만 재학습
    # ------------------------------
    model.fit(X_train_sel, y_train)

    # ------------------------------
    # (D) 예측 및 평가
    # ------------------------------
    y_proba = model.predict_proba(X_val_sel)[:, 1]
    threshold = 0.25   # recall 중심 threshold
    y_pred = (y_proba >= threshold).astype(int)

    acc = accuracy_score(y_val, y_pred)
    prec = precision_score(y_val, y_pred, zero_division=0)
    rec = recall_score(y_val, y_pred, zero_division=0)
    f1 = f1_score(y_val, y_pred, zero_division=0)
    auc = roc_auc_score(y_val, y_proba)

    results.append([fold + 1, acc, prec, rec, f1, auc])

# ==========================================
# 5️⃣ 교차검증 성능 요약
# ==========================================
cv_results = pd.DataFrame(results, columns=['Fold', 'Accuracy', 'Precision', 'Recall', 'F1', 'ROC-AUC'])

print("\n📊 교차검증별 성능:\n", cv_results)
print("\n✅ 교차검증 평균 성능")
print(f"Accuracy: {cv_results['Accuracy'].mean():.4f}")
print(f"Precision: {cv_results['Precision'].mean():.4f}")
print(f"Recall: {cv_results['Recall'].mean():.4f}")
print(f"F1-score: {cv_results['F1'].mean():.4f}")
print(f"ROC-AUC: {cv_results['ROC-AUC'].mean():.4f}")


언더샘플링 후 데이터 크기: (3750, 81)
폐업여부
0    3125
1     625
Name: count, dtype: int64

===== Fold 1 =====

===== Fold 2 =====

===== Fold 3 =====

===== Fold 4 =====

===== Fold 5 =====

📊 교차검증별 성능:
    Fold  Accuracy  Precision    Recall        F1   ROC-AUC
0     1  0.724000   0.322581  0.538462  0.403458  0.647792
1     2  0.686667   0.204301  0.304000  0.244373  0.627251
2     3  0.685333   0.267782  0.512000  0.351648  0.619251
3     4  0.676000   0.264000  0.528000  0.352000  0.691277
4     5  0.694667   0.226131  0.375000  0.282132  0.639683

✅ 교차검증 평균 성능
Accuracy: 0.6933
Precision: 0.2570
Recall: 0.4515
F1-score: 0.3267
ROC-AUC: 0.6451


# 5개월 데이터 선정

불균형 데이터 평가 지표로 Recall값을 중점으로 보았다. 

2개월 - 0.4125
3개월 - 0.4061
4개월 - 0.2290
5개월 - 0.4515

5개월 데이터가 가장 모델을 잘 표현한다고 판단 후 선정.

In [41]:
import joblib

# 학습 끝난 모델 저장
joblib.dump(model, "final_model.pkl")

# 불러오기
model = joblib.load("final_model.pkl")


# 챗봇 구현

제미나이 API를 활용하여 LLM 챗봇 구현 -> 점주의 입장에서 편하고 용이하게 활용 가능.

In [ ]:
# 1. 라이브러리 설치 (Colab에서는 매번 실행해줘야 합니다)
!pip install -q google-generativeai

# 2. 파이썬 코드
import google.generativeai as genai

# --- 중요! ---
# 본인의 API 키를 따옴표("") 안에 정확하게 붙여넣으세요.
# --------------------------------------------------------------------------
API_KEY = "api"
# --------------------------------------------------------------------------

try:
    genai.configure(api_key=API_KEY)
    model = genai.GenerativeModel('gemini-2.5-pro') # Colab에서는 최신 모델도 잘 작동합니다.

    print("--- Gemini AI 챗봇 ---")
    print("무엇이든 물어보세요. 대화를 종료하려면 'exit' 또는 'quit'를 입력하세요.")
    
    while True:
        user_prompt = input("\n나의 질문: ")
        if user_prompt.lower() in ['exit', 'quit']:
            print("프로그램을 종료합니다.")
            break

        print("답변을 생성 중입니다...")
        response = model.generate_content(user_prompt)
        print("\nGemini 답변:", response.text)

except Exception as e:
    print(f"\n[오류 발생] 프로그램을 실행할 수 없습니다.")
    print(f"오류 내용: {e}")

--- Gemini AI 챗봇 ---
무엇이든 물어보세요. 대화를 종료하려면 'exit' 또는 'quit'를 입력하세요.


In [ ]:
# ==========================================
# 1️⃣ 기본 라이브러리
# ==========================================
import pandas as pd
import joblib
import shap
import google.generativeai as genai

# ==========================================
# 2️⃣ Gemini API 설정
# ==========================================
API_KEY = "api"
genai.configure(api_key=API_KEY)
model_gemini = genai.GenerativeModel("models/gemini-2.5-pro")


# ==========================================
# 3️⃣ CatBoost 모델 및 데이터 불러오기
# ==========================================
# ⚠️ 아래 경로에 학습이 완료된 모델이 있어야 함
# (없다면 위쪽에서 joblib.dump(model, "final_catboost_model.pkl") 으로 저장)
model = joblib.load("final_catboost_model.pkl")

# 예시 데이터프레임 불러오기
# (실제 데이터셋에 맞게 변경)
df = pd.read_csv("balanced_data.csv")
X = df.drop(columns=["가맹점ID", "기준년월", "폐업여부"])
y = df["폐업여부"]

# ==========================================
# 4️⃣ SHAP 기반 피처 영향도 계산
# ==========================================
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X)

# 분석할 가맹점 (예: 첫 번째 행)
sample = X.iloc[[0]]
store_id = df.iloc[0]["가맹점ID"]

sample_idx = 0
sample = X.iloc[[sample_idx]]
sample_shap = shap_values[sample_idx]
# 상위 5개 변수 추출
feature_importance = sorted(
    zip(X.columns, sample_shap),
    key=lambda x: abs(x[1]),
    reverse=True
)[:5]

# 텍스트 변환
feature_text = "\n".join([f"- {f[0]}: 영향도 {f[1]:.4f}" for f in feature_importance])

print(feature_text)

In [ ]:
import pandas as pd

# ==========================================
# 1️⃣ 변수 중요도 계산 (길이 자동 동기화)
# ==========================================
fi = pd.DataFrame({
    'feature': model.feature_names_,                     # ← 모델 내부에 저장된 실제 피처 이름
    'importance': model.get_feature_importance()
}).sort_values('importance', ascending=False)

top_features = fi.head(10)['feature'].tolist()  # 상위 10개 변수만 선택

# ==========================================
# 2️⃣ 폐업 / 비폐업 점포 분리
# ==========================================
df_closed = df[df['폐업여부'] == 1].copy()
df_open = df[df['폐업여부'] == 0].copy()

# ==========================================
# 3️⃣ 폐업 점포 임계값 요약 (평균, 중앙값, 분위수)
# ==========================================
thresholds = df_closed[top_features].agg(['mean', 'median'])
thresholds.loc['q25'] = df_closed[top_features].quantile(0.25)
thresholds.loc['q75'] = df_closed[top_features].quantile(0.75)

print("📊 폐업 점포의 주요 변수 임계값 요약:")
print(thresholds.T)

# ==========================================
# 4️⃣ 폐업 vs 비폐업 평균 비교
# ==========================================
compare = pd.concat([
    df_closed[top_features].mean().rename("폐업_mean"),
    df_open[top_features].mean().rename("비폐업_mean")
], axis=1)

compare["gap"] = compare["폐업_mean"] - compare["비폐업_mean"]
compare = compare.sort_values("gap", ascending=False)

print("\n📉 폐업 vs 비폐업 평균 비교 (gap 기준 정렬):")
print(compare)

# ==========================================
# 5️⃣ 결과 해석 팁
# ==========================================
# gap (+) → 폐업 점포에서 평균적으로 더 높은 값 → 이 변수가 높을수록 위험↑
# gap (-) → 폐업 점포에서 평균적으로 더 낮은 값 → 이 변수가 낮을수록 위험↑


In [ ]:
# ==========================================
# 🧩 전처리 함수 정의 (훈련/예측 공용)
# ==========================================
import pandas as pd
import numpy as np

def preprocess_features(df: pd.DataFrame) -> pd.DataFrame:
    """
    balanced_data.csv에서 사용한 주요 파생 변수 생성 로직을
    store_features.csv에도 동일하게 적용하는 함수
    """
    # ✅ 1. 운영개월수구간 생성
    if "운영개월수" in df.columns and "운영개월수구간" not in df.columns:
        df["운영개월수구간"] = pd.cut(
            df["운영개월수"],
            bins=[0, 12, 24, 36, 48, 60, np.inf],
            labels=[1, 2, 3, 4, 5, 6]
        ).astype(float)

    # ✅ 2. 매출변화율, 매출변화율_3개월평균 등 계산
    if all(col in df.columns for col in ["매출금액_1개월전", "매출금액_2개월전", "매출금액_3개월전"]):
        df["매출변화율"] = (
            (df["매출금액_1개월전"] - df["매출금액_2개월전"]) / (df["매출금액_2개월전"] + 1e-6)
        )
        df["매출변화율_3개월평균"] = (
            (df["매출금액_1개월전"] + df["매출금액_2개월전"] + df["매출금액_3개월전"]) / 3
        ) / (df["매출금액_3개월전"] + 1e-6) - 1

    # ✅ 3. 성별/연령별 고객 비중 이동평균 (없으면 건너뜀)
    for col in df.columns:
        if ("고객 비중" in col) and ("_4개월평균" not in col):
            new_col = f"{col}_4개월평균"
            df[new_col] = (
                df.groupby("가맹점ID")[col].transform(lambda x: x.rolling(4, 1).mean())
                if "가맹점ID" in df.columns else df[col]
            )

    # ✅ 4. 비율형 지표의 결측치 보정
    ratio_cols = [c for c in df.columns if "비율" in c or "순위" in c]
    df[ratio_cols] = df[ratio_cols].fillna(df[ratio_cols].median())

    # ✅ 5. 결측치 전체 보정
    df = df.fillna(0)

    return df


# 챗봇 성능 보안

챗봇이 단순히 +- 값만으로 판단하지 않도록 성능을 보안.

In [ ]:
# ==========================================
# 0️⃣ 라이브러리
# ==========================================
import pandas as pd
import numpy as np
import joblib
import shap
import google.generativeai as genai


# ==========================================
# 1️⃣ 환경 설정 (API 키, 모델 로드)
# ==========================================
GEMINI_API_KEY = "api"
genai.configure(api_key=GEMINI_API_KEY)
llm = genai.GenerativeModel("models/gemini-2.5-pro")

# CatBoost 모델 로드
model = joblib.load("final_catboost_model.pkl")

# 모델이 학습 시 사용한 정확한 feature 이름 가져오기
trained_features = model.feature_names_



# ==========================================
# 2️⃣ 예측용 데이터 준비
# ==========================================
df = pd.read_csv("store_features.csv")

# 🔹 (여기 추가!) 전처리 함수 적용
df = preprocess_features(df)

# 🔹 모델 피처 순서 맞추기
X = df.reindex(columns=trained_features, fill_value=0)



# ==========================================
# 3️⃣ 예측 및 위험 등급 계산
# ==========================================
pred_prob = model.predict_proba(X)[:, 1][0]

if pred_prob >= 0.30:
    risk_level = "높음"
elif pred_prob >= 0.15:
    risk_level = "중간"
else:
    risk_level = "낮음"


# ==========================================
# 4️⃣ SHAP 영향도 계산
# ==========================================
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X)

row_idx = 0
sample_shap = shap_values[row_idx]

# 상위 5개 중요 변수 추출
top_k = 10
feature_importance = sorted(
    zip(trained_features, sample_shap),
    key=lambda x: abs(x[1]),
    reverse=True
)[:top_k]

# 영향도 텍스트 생성
shap_explain_lines = []
for fname, sval in feature_importance:
    direction = "위험↑ 요인" if sval > 0 else "안정↑ 요인"
    shap_explain_lines.append(f"- {fname}: 영향도 {sval:.4f} ({direction})")

feature_text = "\n".join(shap_explain_lines)

print("📊 주요 영향 변수:")
print(feature_text)


# ==========================================
# 5️⃣ 폐업점포 기준 임계값 계산 (개선 버전)
# ==========================================
df_train = pd.read_csv("balanced_data.csv")
df_closed = df_train[df_train["폐업여부"] == 1]

# 상위 피처 추출
top_features = [f[0] for f in feature_importance]

# ✅ 존재하는 컬럼만 선택 (부분 일치 허용)
available_features = []
for f in top_features:
    matches = [col for col in df_closed.columns if f.replace(" ", "") in col.replace(" ", "")]
    if matches:
        available_features.append(matches[0])

# ✅ 임계값 계산
thresholds = df_closed[available_features].agg(["mean", "median"])
thresholds.loc["q25"] = df_closed[available_features].quantile(0.25)
thresholds.loc["q75"] = df_closed[available_features].quantile(0.75)

# ✅ SHAP 방향성 기반 위험 방향 설정
feature_direction = {}
for fname, sval in feature_importance:
    if sval > 0:
        feature_direction[fname] = "high"  # 값이 높을수록 위험
    else:
        feature_direction[fname] = "low"   # 값이 낮을수록 위험

# ✅ 현재 매장 데이터에서 동일 컬럼 추출
current_values = {}
for f in available_features:
    match = [col for col in df.columns if f.replace(" ", "") in col.replace(" ", "")]
    if match:
        current_values[f] = df[match[0]].iloc[0]

# ✅ 비교 및 해석 텍스트 생성
threshold_text = ""
for col in available_features:
    if col not in current_values:
        threshold_text += f"- {col}: ❌ 현재 데이터에 해당 컬럼 없음\n"
        continue

    val = current_values[col]
    avg = thresholds.loc["mean", col]
    q25 = thresholds.loc["q25", col]
    q75 = thresholds.loc["q75", col]
    direction = feature_direction.get(col, "unknown")

    # ⚙️ 위험 판단 로직 (방향성 반영)
    if direction == "high":
        status = "⚠️ 위험구간" if val >= q75 else "✅ 안정범위"
    elif direction == "low":
        status = "⚠️ 위험구간" if val <= q25 else "✅ 안정범위"
    else:
        status = "⚪ 판단불가"

    threshold_text += (
        f"- {col}: 현재={val:.3f}, 폐업점 평균={avg:.3f}, 방향={direction} ({status})\n"
    )

print("\n📏 주요 변수 임계값 비교 (방향성 반영):")
print(threshold_text)

In [ ]:
# ==========================================
# 6️⃣ Gemini 리포트 생성
# ==========================================
prompt = f"""
당신은 프랜차이즈 본사의 데이터 기반 경영 컨설턴트입니다.

이 가맹점(ID: {store_id})의 폐업 위험 예측 확률은 {pred_prob:.2%} (위험 수준: {risk_level}) 입니다.
모델이 판단한 주요 영향 요인은 다음과 같습니다:

{feature_text}

또한 아래는 폐업 점포들의 평균 임계값과 비교한 현재 매장의 실제 값입니다.
(⚠️은 위험구간, ✅은 안정구간을 의미)

{threshold_text}

이 정보를 바탕으로 다음을 작성하세요:
1. 폐업 위험이 {risk_level}으로 나타난 주요 데이터 기반 이유
2. 임계값을 기준으로 현재 가맹점이 주의해야 할 지표 2가지
3. 3개월 내 개선 가능한 핵심 실천 전략 2가지
4. 같은 동 내 같은 업종 평균 대비 강점과 약점
5. 사장님이 이해하기 쉬운 언어로, 위협적이지 않게 코칭하듯 작성

반드시 한국어로, 친절하고 실용적으로 작성하세요.
"""

response = llm.generate_content(prompt)

print("\n📋 Gemini 리포트 요약\n")
print(response.text)

In [ ]:
# 1. 라이브러리 설치 (Colab에서는 매번 실행해줘야 합니다)
!pip install -q google-generativeai

# 2. 파이썬 코드
import google.generativeai as genai

# --- 중요! ---
# 본인의 API 키를 따옴표("") 안에 정확하게 붙여넣으세요.
# --------------------------------------------------------------------------
API_KEY = "api"
# --------------------------------------------------------------------------

try:
    genai.configure(api_key=API_KEY)
    model = genai.GenerativeModel('gemini-2.5-pro') # Colab에서는 최신 모델도 잘 작동합니다.

    print("--- Gemini AI 챗봇 ---")
    print("무엇이든 물어보세요. 대화를 종료하려면 'exit' 또는 'quit'를 입력하세요.")
    
    while True:
        user_prompt = input("\n나의 질문: ")
        if user_prompt.lower() in ['exit', 'quit']:
            print("프로그램을 종료합니다.")
            break

        print("답변을 생성 중입니다...")
        response = model.generate_content(user_prompt)
        print("\nGemini 답변:", response.text)

except Exception as e:
    print(f"\n[오류 발생] 프로그램을 실행할 수 없습니다.")
    print(f"오류 내용: {e}")